TECHNICAL_DETAILS.md

```diff
@@ -0,0 +1,274 @@
+
+![Savings Graph](https://cdn.cloudprowess.com/images/autospotting-savings.png)
+
+## Best Practices ##
+
+These recommendations apply for most cloud environments, but they become
+especially important when using more volatile spot instances.
+
+- **Set a non-zero grace period on the AutoScaling group**
+  - in order to attach spot instances only after they are fully configured.
+  - otherwise they may be attached prematurely before being ready.
+  - they may also be terminated after failing load balancer health checks.
+
+- **Check your instance storage and block device mapping configuration**
+  - this may become an issue if you use instances which have ephemeral instance
+    storage, often the case on previous instance types.
+  - you should only specify ephemeral instance store in the on-demand launch
+    configuration if you do make use of it by mounting it on the filesystem.
+  - the replacement algorithm tries to give you instances with as much instance
+    storage as your original instances, since it can't tell if you did mount it.
+  - this adds more constraints on the algorithm, so it reduces the number of
+    compatible instance types it can use for launching spot instances.
+  - this is fine if you actually use that instance storage, but it is reducing
+    your options if you don't actually use it, so it may more often fail to get
+    spot instances and fall back to on-demand capacity.
+
+- **Don't keep state on instances**
+  - You should delegate all your state to external services, AWS has a wide
+    offering of stateful services which allow your instances to become
+    stateless.
+    - Databases: RDS, DynamoDB
+    - Caches: ElastiCache
+    - Storage: S3, EFS
+    - Queues: SQS
+  - Don't attach EBS volumes to individual instances, try to use EFS instead.
+
+- **Handle the spot instance termination signal**
+  - See the next section for more detailed instructions.
+
+## Spot termination notifications ##
+
+AWS
+[notifies](https://aws.amazon.com/blogs/aws/new-ec2-spot-instance-termination-notices/)
+your spot instances when they are about to be terminated by setting a dedicated
+metadata field, so you can make use of that information to save whatever
+temporary state you may still have on your running spot instances or to
+gracoiusly remove them from the group.
+
+There are existing third party tools which implement such a termination
+notification handler, such as [seespot](https://github.com/acksin/seespot). This
+will need to be integrated into your user_data script, for more details you can
+read see the seespot tool's documentation.
+
+### Instances behind an ELB ###
+
+Instances behind an ELB can be graciously
+[removed](https://aws.amazon.com/blogs/aws/elb-connection-draining-remove-instances-from-service-with-care/)
+from the load balancer without losing connections. You should enable the
+connection draining feature, and then you just need to append this snippet to
```

Comment:
I don't have it, but I will mention that it could be done, hopefully someone will come up with more details.